doc/developer/design/20230306_persist_mfp_pushdown.md

```diff
@@ -0,0 +1,272 @@
+- Even in the initial version of this feature, we have customers that would like to push down filters that use the `->` operator on `jsonb` objects/maps.
+
+As a result, we end up with the full stats being recursive and looking something like this (simplified here for clarity):
+
+```rust
+struct LowerAndUpperBounds<T>{lower: T, upper: T};
+struct JsonbStats{
+    json_null_count: usize,
+    bools: Option<LowerAndUpperBounds<bool>>,
+    numerics: Option<LowerAndUpperBounds<Numeric>>,
+    strings: Option<LowerAndUpperBounds<String>>,
+    lists: Option<Vec<JsonbStats>>,
+    maps: Option<BTreeMap<String, JsonbStats>>,
+};
+```
+
+This, of course, can explode to unbounded size given a sufficiently adversarial user. See the next section.
+
+## Statistics trimming
+
+Because statistics are stored in metadata, it is desireable to keep a bound on the size of them. Even if we guarantee that each individual column's statistics are of bounded size, we could have any number of columns. If we had perfect knowledge at statistics calculation time of which columns would be queried, this is much easier, but one of the benefits of this feature is that it works without configuration and also enables exploratory queries. So, ideally we don't require the user to declare this up front.
+
+There are a number of options here, including machine learning models, but simple heuristics are understandable and can get us surprisingly far. It is difficult to predict exactly what the heuristics will be, but an initial thought is something like the following:
+
+- First, a budget (in bytes) is calculated, possibly as a function of how many updates are in the part. We guarantee that the resulting statistics fit into this budget.
+- Second, anything obviously unnecessary is trimmed. For example, any real query filtering on a json column is likely to have a cast in the expression tree. This means that many (all?) json stats computed on a column containing more than one Datum type (e.g. both `Numeric` and `String`) would output an error from MFP evaluation for something in the part. As a result, the stats would never allow us to filter this part and thus can be trimmed.
+- Third, we can adjust truncation of stats over varlen columns. For example, a bound of `a<huge string>..=y<huge string>` can likely be shortened to `a..=z`. This does make the bounds less tight, which increases false positives.
+- Finally, we can trim stats for `Record`, `List`, or `Jsonb` elements or even for entire columns. Note that if we do this, we almost certainly want some way for the customer to be able to hint that a certain column is critical and should never have trimmed stats (possibly even something as simple as a LaunchDarkly flag).
```

Comment:
is it possible to compute stats up to some budget, rather than needing to trim away computed statistics?